### PR TITLE
Cache test corpus in testdata

### PR DESCRIPTION
**Pull request by hwh33**
This downloads/updates the test corpus when the tests are run.  You can use the `-short` flag to skip the update process and keep tests from running against the larger corpus (they'll still run against anything else in `testdata`).  If you are offline or if there is some other issue checking for updates, you'll get a message on stderr like
```
failed to update test corpus: HEAD failed: Head "https://meta-scrubber-test-corpus.s3.us-west-1.amazonaws.com/exif-image-corpus.tar.gz": dial tcp: lookup meta-scrubber-test-corpus.s3.us-west-1.amazonaws.com: no such host
```
but the tests will still run against whatever is already downloaded.  When you update the corpus and run the tests, you'll see output on stderr like
```
ETag changed; downloading new test corpus
Successfully updated test corpus
```
Otherwise, there is no additional output; the tests just run like normal.

TL;DR, you can just `git clone` and `go test` to test against the corpus.  At any time, you can run `go test -short` to skip all the long-running stuff.

What do you think @max-b?

### CODE REVIEW

**Comment by hwh33** [comment]
Also, came across this great image in the corpus:

image

**Comment by max-b** [comment]
> you'll get a message on stderr like

Just out of curiosity, why `stderr` (and not use `t.Log`) here?

**Comment by max-b** [comment]
Yeah in general I definitely like the simplicity of being able to just run `git clone` and then `go test`

I wonder if we could change it a bit so that the flag is for whether or not to attempt to download the corpus? I guess we could also have a `short` flag which skips the corpus directory if you want shorter tests (or maybe a flag to check the validity of a single image?)

Would you mind updating the circleci config to use your new approach here? I _think_ for caching we'd still need to use the curl command to check the etag manually in the beginning, but I'm wondering if you have a better idea there.


Thanks!

**Comment by hwh33** [comment]
> Just out of curiosity, why stderr (and not use t.Log) here?

I would have used `t.Log`, but I didn't have access to it in the `init` function.  I could instead have every test call some kind of `updateCorpusIfNecessary` function and pass the `testing.T` instance into that.  But then we'd need to remember to add that call to every new test and I didn't like that.  So I just tried to really limit the messages that go to stderr.

The `-short` flag is actually defined by the Go tool.  Tests can call [testing.Short](https://golang.org/pkg/testing/#Short) to check whether the flag is set.  The reason I didn't separate out updating the corpus and testing against the corpus is that I didn't see a use case.  On my machine, the tests alone take about 2 minutes.  The HEAD request and ETag check is quick (~0.5s for me), so when the corpus is unchanged, the difference is negligible.  If someone cares enough to wait 2 minutes for the tests, wouldn't they also want to make sure they're testing against the current corpus?  That was my take anyway.  If you think there is a use case for a separate "skip corpus update" flag, I'm happy to add that.

> Would you mind updating the circleci config to use your new approach here?

Ooo yes, good catch!  Will do.

**Comment by max-b** [comment]
> The -short flag is actually defined by the Go tool. Tests can call testing.Short to check whether the flag is set. The reason I didn't separate out updating the corpus and testing against the corpus is that I didn't see a use case. On my machine, the tests alone take about 2 minutes. The HEAD request and ETag check is quick (~0.5s for me), so when the corpus is unchanged, the difference is negligible. If someone cares enough to wait 2 minutes for the tests, wouldn't they also want to make sure they're testing against the current corpus? That was my take anyway. If you think there is a use case for a separate "skip corpus update" flag, I'm happy to add that.

Ok yeah that makes sense - I'm always happy to stick w/ idioms

File: go.mod [hunk 1]
```diff
@@ -3,7 +3,6 @@ module github.com/getlantern/meta-scrubber
 go 1.14
 
 require (
-	github.com/dsoprea/go-jpeg-image-structure v0.0.0-20200615034914-d40a386309d2
```
Comment (hwh33) [inline go.mod:6]: `go mod tidy` did this

File: go.sum [hunk 2]
```diff
@@ -46,6 +38,7 @@ golang.org/x/net v0.0.0-20200520182314-0ba52f642ac2/go.mod h1:qpuaurCH72eLCgpAm/
 golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
 golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
 golang.org/x/text v0.3.0/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
+gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405 h1:yhCVgyC4o1eVCa2tZl7eS0r+SDo693bJlVdllGtEeKM=
```
Comment (hwh33) [inline go.sum:41]: `go mod tidy` made the changes to this file

File: scrubber_test.go [hunk 3]
```diff
@@ -114,3 +150,86 @@ func TestScrubbing(t *testing.T) {
...
+	}
+	defer headResp.Body.Close()
+	if headResp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad response to HEAD request: %s", headResp.Status)
+	}
+	etag := headResp.Header.Get("ETag")
+	if etag == "" {
+		return errors.New("no ETAG in HEAD response")
+	}
+	if etag == strings.TrimSpace(string(currentETag)) {
```
Comment (hwh33) [inline scrubber_test.go:171]: This is a bit of a hack.  In CircleCI, the ETag file will contain a trailing newline so we trim the file of space characters just in case.

**Comment by hwh33** [comment]
Okay, CircleCI is now caching the corpus in a manner compatible with how the tests cache the corpus.  It just saves it to the same location and writes the ETag to the expected location as well.  It's a little hacky I suppose, but it should hold up okay.  I also added a note in `scrubber_test.go` about the need for CircleCI to agree with the location of the corpus / ETag file.

**Comment by hwh33** [comment]
FYI, I noticed [this issue](https://github.com/getlantern/meta-scrubber/issues/4) while debugging CI stuff.

**Comment by max-b** [comment]
> Okay, CircleCI is now caching the corpus in a manner compatible with how the tests cache the corpus. It just saves it to the same location and writes the ETag to the expected location as well. It's a little hacky I suppose, but it should hold up okay. I also added a note in `scrubber_test.go` about the need for CircleCI to agree with the location of the corpus / ETag file.

I _think_ that circleCI is re-downloading the corpus data every time? I'm wondering if we can maybe just scrap the `curl` parts from the circleCI piece entirely. I can try and throw something together to demo what I mean...

**Comment by hwh33** [comment]
> I think that circleCI is re-downloading the corpus data every time?

I saw that there was already logic to cache the corpus, so I just assumed that was working.  Is that not the case?  What I verified is that the corpus is not re-downloaded in the test stage (we don't see the "downloading new test corpus" message).

**Comment by hwh33** [comment]
As far as scrapping the curl parts, my assumption was that the repo folder, and therefore the testdata folder, are ephemeral.  CircleCI may be building and testing multiple commits at once, so they can't share the same repo folder, right?  I think that necessitates that CircleCI manage its own corpus caching.  Am I missing something?

**Comment by hwh33** [comment]
Okay, it looks like caching the corpus in CircleCi isn't working at the moment.  You can see this if you go to the master branch in CircleCI and just re-run the latest build (I'm assuming that should use the cached corpus).  I can look into it

File: .circleci/config.yml [hunk 4]
```diff
@@ -9,7 +9,7 @@ jobs:
       - run:
           name: Get test corpus etag
           command: |
-            curl https://meta-scrubber-test-corpus.s3.us-west-1.amazonaws.com/exif-image-corpus.tar.gz -I | grep -Fi ETag | sed 's/ETag: //g' > /tmp/exif-image-corpus-etag
+            curl https://meta-scrubber-test-corpus.s3.us-west-1.amazonaws.com/exif-image-corpus.tar.gz -I | grep -Fi ETag | sed 's/ETag: //g' | sed 's/"//g' > /tmp/exif-image-corpus-etag
```
Comment (hwh33) [inline .circleci/config.yml]: Oh the test code actually expects quotes.  We can change that if we need to though

Comment (max-b) [reply .circleci/config.yml]: ooooh yeah I just realized that too - the thing that was actually causing the problems were the newlines at the end of that etag file. I believe I just finally got it working :grin:

Comment (max-b) [reply .circleci/config.yml]: https://app.circleci.com/pipelines/github/getlantern/meta-scrubber/48/workflows/a391024d-3ab7-4824-b5a0-21377b2b585c/jobs/57/parallel-runs/0/steps/0-104

Comment (hwh33) [reply .circleci/config.yml]: Hey nice!  Looks good

**Review by max-b — APPROVED** [review]
Great thanks @hwh33 !!
